### PR TITLE
Issue 771233: Support udp_sender option from libvma configuration

### PR DESCRIPTION
**Pull request by ashanably**
Support udp_sender and use it instead of udp_connect

Signed-off-by: ashanably <ashanably@asaltech.com>

### CODE REVIEW

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

File: src/vma/sock/sockinfo_udp.cpp [hunk 1]
```diff
@@ -582,7 +582,7 @@ int sockinfo_udp::connect(const struct sockaddr *__to, socklen_t __tolen)
 		si_udp_logdbg("bound to %s", m_bound.to_str());
 		in_port_t src_port = m_bound.get_in_port();
 
-		if (TRANS_VMA != find_target_family(ROLE_UDP_CONNECT, m_connected.get_p_sa(), m_bound.get_p_sa())) {
+		if (TRANS_VMA != find_target_family(ROLE_UDP_SENDER, m_connected.get_p_sa(), m_bound.get_p_sa())) {
```
Comment (OphirMunk) [inline src/vma/sock/sockinfo_udp.cpp]: I find no reason for calling ROLE_UDP_SENDER under sockinfo_udp::connect()
The original role (ROLE_UDP_CONNECT) seems correct

File: src/vma/sock/sockinfo_udp.cpp [hunk 2]
```diff
@@ -1581,9 +1582,14 @@ ssize_t sockinfo_udp::tx(const tx_call_t call_type, const struct iovec* p_iov, c
 			si_udp_logdbg("MSG_OOB not supported in UDP (tx-ing to os)");
 			goto tx_packet_to_os;
 		}
-
+		
 		sock_addr dst((struct sockaddr*)__dst);
-
+
```
Comment (OphirMunk) [inline src/vma/sock/sockinfo_udp.cpp]: 1. Adding this code (find_target_family()...) here will hurt latency in UDP fast path.
2. It seems that ROLE_UDP_SENDER is already supported in code.
Checking of ROLE_UDP_SENDER already exists in dst_entry_udp.cpp
```code
transport_t dst_entry_udp::get_transport(sockaddr_in to)
{
	return  __vma_match_udp_sender(TRANS_VMA, mce_sys.app_id, (sockaddr *)(&to), sizeof to);
}
```

get_transport() method is a nested call - the result of dst_entry_udp::slow_send()

```code2
ssize_t dst_entry_udp::slow_send(const iovec* p_iov, size_t sz_iov, bool b_blocked /*= true*/, bool is_rexmit /*= false*/, int flags /*= 0*/, socket_fd_api* sock /*= 0*/, tx_call_t call_type /*= 0*/)
{
	NOT_IN_USE(is_rexmit);

	ssize_t ret_val = 0;

	dst_udp_logdbg("In slow send");

	prepare_to_send();  // <========================== m_b_force_os is set to true

	if (m_b_force_os || !m_b_is_offloaded) {
		struct sockaddr_in to_saddr;
		to_saddr.sin_port = m_dst_port;
		to_saddr.sin_addr.s_addr = m_dst_ip.get_in_addr();
		to_saddr.sin_family = AF_INET;
		dst_udp_logdbg("Calling to tx_os");
                //   <=============================== calling OS on next line
		ret_val = sock->tx_os(call_type, p_iov, sz_iov, flags, (const struct sockaddr*)&to_saddr, sizeof(struct sockaddr_in)); 
	}
        ...............
```

Please update libvma.conf file such that by default DNS protocol is not offloaded. Use ROLE_UDP_SENDER

File: src/vma/sock/sockinfo_udp.cpp [hunk 3]
```diff
@@ -1581,9 +1582,14 @@ ssize_t sockinfo_udp::tx(const tx_call_t call_type, const struct iovec* p_iov, c
...
-
+		
 		sock_addr dst((struct sockaddr*)__dst);
-
+	
+		if (TRANS_VMA != find_target_family(ROLE_UDP_SENDER, dst.get_p_sa(), m_bound.get_p_sa())) {
+			setPassthrough();
+			goto tx_packet_to_os;
+		}	
+
```
Comment (OphirMunk) [inline src/vma/sock/sockinfo_udp.cpp]: If there is a configuration rule on DNS port - why doing setPassthrough() to the socket? 
Can't it be there is other offloaded traffic on this socket?

**Comment by OphirMunk** [comment]
ROLE_UDP_SENDER is supported in code

**Pull request closed without merging**